### PR TITLE
[dtensor][8/N] switch DeviceMesh to use numpy array for devices

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92126
* __->__ #92931
* #90735
* #90734
* #90733

Switching from torch.Tensor to numpy array to avoid possible
interactions with tracing subsystems

Differential Revision: [D42876247](https://our.internmc.facebook.com/intern/diff/D42876247)

### CODE REVIEW

File: torch/distributed/_tensor/device_mesh.py [hunk 1]
```diff
@@ -38,19 +40,6 @@ def set_global_device_mesh(mesh: Optional["DeviceMesh"]) -> None:
...
-T = TypeVar("T")
-_L = Union[T, Sequence[T]]
-NDIntList = _L[_L[_L[_L[_L[_L[_L[int]]]]]]]
-
-MeshExprT = Union[
-    torch.Tensor,
-    NDIntList,
-]
-
-
```
Comment (aazzolini) [inline torch/distributed/_tensor/device_mesh.py:53]: Thanks this was a bit weird :)

**Review by aazzolini — APPROVED** [review]
Thanks! One suggestion for unit testing would be to create a DeviceMesh in FakeMode to reproduce the issue that I had! Or maybe create a DeviceMesh inside of a simple function and make_fx on it?

**Comment by wanchaol** [comment]
> Thanks! One suggestion for unit testing would be to create a DeviceMesh in FakeMode to reproduce the issue that I had! Or maybe create a DeviceMesh inside of a simple function and make_fx on it?

Thanks @aazzolini! Just added a test with the FakeTensorMode, it works as expected. However if we want to fully avoid this issue, we might want to stop supporting DeviceMesh taking tensor as input for the mesh field, and use either numpy array or n-d list instead. Let me know if you want to go with that direction :)

**Comment by XilunWu** [comment]
> > Thanks! One suggestion for unit testing would be to create a DeviceMesh in FakeMode to reproduce the issue that I had! Or maybe create a DeviceMesh inside of a simple function and make_fx on it?
> 
> Thanks @aazzolini! Just added a test with the FakeTensorMode, it works as expected. However if we want to fully avoid this issue, we might want to stop supporting DeviceMesh taking tensor as input for the mesh field, and use either numpy array or n-d list instead. Let me know if you want to go with that direction :)

Would allowing passing Tensor as argument but converting it to numpy array or n-d list in `DeviceMesh._init_` avoid the tracing issue?

**Comment by wanchaol** [comment]
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42876247).

**Comment by wanchaol** [comment]
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D42876247).

**Pull request closed without merging**